https://issues.redhat.com/browse/ACM-5430 deploy cluster cli 2.8 (#5153)
open-cluster-management/rhacm-docs: clusters/cluster_lifecycle/create_cluster_on_prem.adoc

[#on-prem-creating-your-cluster-with-the-cli-namespace]
=== Create the namespace

You need a namespace for your resources. It is more convenient to keep all of the resources in a shared namespace. This example uses `sample-namespace` for the name of the namespace, but you can use any name except `assisted-installer` Create a namepace by creating and applying the following file:

[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
  name: sample-namespace
----

[#on-prem-creating-your-cluster-with-the-cli-pull-secret]
=== Add the pull secret to the namespace

Add your link:https://console.redhat.com/openshift/install/pull-secret[pull secret] to your namespace by creating and applying the following custom resource:

[source,yaml]
----
apiVersion: v1
kind: Secret
type: kubernetes.io/dockerconfigjson
metadata:
  name: pull-secret
  namespace: sample-namespace
stringData:

>   .dockerconfigjson: '{"auths":{"cloud.openshift.com":{"auth":"your-secret-here","email":"user@example.com"}}}'

Reviewer: This isn't quite correct. This needs to contain more than just `cloud.openshift.com` auth.

For example, I'd expect `quay.io` and `registry.redhat.io` auth as well in here.
Maybe it's best to just remove this sample text in favor of a single placeholder? Something like:

```
  .dockerconfigjson: 'your-pull-secret-json'
```

Author: Should we include a list of what it might include, or is it clear enough with just the generic 'your-pull-secret-json'?

Reviewer: I think the substitute is enough. I'm also not sure what all the different bits of this might be and those might change over time.

apiVersion: hive.openshift.io/v1
kind: ClusterImageSet
metadata:
   name: openshift-v4.8.0
spec:
   releaseImage: quay.io/openshift-release-dev/ocp-release:4.8.0-rc.0-x86_64  # That version may change over time
----

[#on-prem-creating-your-cluster-with-the-cli-clusterdeployment]
=== Create the ClusterDeployment custom resource

The `ClusterDeployment` custom resource definition is an API that controls the lifecycle of the cluster. It references the `AgentClusterInstall` custom resource in the `spec.ClusterInstallRef` setting which defines the cluster parameters. 

Create and apply a `ClusterDeployment` custom resource based on the following example:

[source,yaml]
----
apiVersion: hive.openshift.io/v1
kind: ClusterDeployment
metadata:
   name: single-node
   namespace: demo-worker4
spec:
   baseDomain: hive.example.com
   clusterInstallRef:
      group: extensions.hive.openshift.io
      kind: AgentClusterInstall
      name: test-agent-cluster-install # Use the same name for AgentClusterInstall
      version: v1beta1
   clusterName: test-cluster
   controlPlaneConfig:
      servingCertificates: {}
   platform:
      agentBareMetal:
         agentSelector:
            matchLabels:

>                bla: aaa

Reviewer: @nmagnezi @filanov do you know if we need this?

Seems strange to put this in official documentation.

Reviewer 2: Agree, suggesting to put something more meaningful, for example location or something like that

   pullSecretRef:

>          name: pull-secret  # Use the pull secret name mentioned in Cluster Prerequisites step 2

Reviewer: Indentation looks off here.

          next-hop-interface: eth1
          table-id: 254
        - destination: 0.0.0.0/0
          next-hop-address: 192.168.140.1
          next-hop-interface: eth1
          table-id: 254
  interfaces:
    - name: "eth0"
      macAddress: "02:00:00:80:12:14"
    - name: "eth1"
      macAddress: "02:00:00:80:12:15"
----

[#on-prem-creating-your-cluster-with-the-cli-infraenv]
=== Create the InfraEnv custom resource

The `InfraEnv` custom resource provides the configuration to create the discovery ISO. Within this custom resource, you identify values for proxy settings, ignition overrides, and specifying `NMState` labels. The `spec.nmStateConfigLabelSelector` in this custom resource references the `NMStateConfig` custom reference, so this custom resource must be created after the `NMStateConfig` custom resource if you included one. 

Create and apply the following custom resource:

[source,yaml]
----
apiVersion: agent-install.openshift.io/v1beta1
kind: InfraEnv
metadata:
  name: myinfraenv
  namespace: demo-worker4
spec:
  clusterRef:
    name: single-node  # Use the above created clusterDeployment resource name and namespace
    namespace: demo-worker4

>   pullSecretRef: # currently ignored by InfraEnv controller

Reviewer: This is not true. This is very much used.

    name: pull-secret
 #sshAuthorizedKey: 'your_pub_key_here' (optional), this key will allow to connect to machines booted from the discovery ISO.
  nmStateConfigLabelSelector:
    matchLabels:
      demo-nmstate-label: some-value # both the label name and value must match the NMStateConfig label section
----

[#on-prem-creating-your-cluster-with-the-cli-boot-host]
=== Boot the host from the discovery image

The remaining steps explain how to boot the host from the discovery ISO image that results from the previous procedures. 

. Download the discovery image from the namespace by running the following command:
+
----
curl --insecure -o image.iso $(kubectl -n sample-namespace get infraenvs.agent-install.openshift.io myinfraenv -o=jsonpath="{.status.isoDownloadURL}")
----

. Move the discovery image to virtual media, a USB drive, or another storage location and boot the host from the discovery image that you downloaded. 

. Approve the `Agent` custom resource and start the installation by running the following command:
+
----
kubectl -n sample-namespace patch agents.agent-install.openshift.io 07e80ea9-200c-4f82-aff4-4932acb773d4 -p '{"spec":{"approved":true}}' --type merge
----
+

> The `Agent` resource is created automatically. It is registered to the cluster and represents a host that can boot from a discovery image. 

Reviewer: ```suggestion
The `Agent` resource is created automatically. It is registered to the cluster and represents a host that has booted from a discovery image. 
```

[#on-prem-creating-your-cluster-with-the-cli-namespace]
=== Create the namespace

You need a namespace for your resources. It is more convenient to keep all of the resources in a shared namespace. This example uses `sample-namespace` for the name of the namespace, but you can use any name except `assisted-installer` Create a namepace by creating and applying the following file:

[source,yaml]
----
apiVersion: v1
kind: Namespace
metadata:
  name: sample-namespace
----

[#on-prem-creating-your-cluster-with-the-cli-pull-secret]
=== Add the pull secret to the namespace

Add your link:https://console.redhat.com/openshift/install/pull-secret[pull secret] to your namespace by creating and applying the following custom resource:

[source,yaml]
----
apiVersion: v1
kind: Secret
type: kubernetes.io/dockerconfigjson
metadata:
  name: pull-secret
  namespace: sample-namespace

> stringData:

Reviewer 3: `stringData:` should just be `data:`

Reviewer 3: Actually nevermind - stringData works here. stringData will take the pull secret as a string, convert to base64, and move to the `data:` field.

  .dockerconfigjson: 'your-pull-secret-json' <1>
----

> <1> Add the name of your pull secret authentication file. For example, this can include a `cloud.openshift.com`, `quay.io`, or `registry.redhat.io` authentication.

Reviewer: This is not a file name, it's the actual pull secret content.

[#on-prem-creating-your-cluster-with-the-cli-cluster-image-set]
=== Generate a ClusterImageSet

Generate a `CustomImageSet` to specify the version of {ocp-short} for your cluster by creating and applying the following custom resource:

[source,yaml]
----
apiVersion: hive.openshift.io/v1
kind: ClusterImageSet
metadata:
  name: openshift-v4.12.0
spec:
  releaseImage: quay.io/openshift-release-dev/ocp-release:4.12.0-rc.0-x86_64  <1>
----

> <1> Your release image might change as you update your cluster. 

Reviewer: This isn't relevant for update after the cluster is installed.

[#on-prem-creating-your-cluster-with-the-cli-clusterdeployment]
=== Create the ClusterDeployment custom resource

The `ClusterDeployment` custom resource definition is an API that controls the lifecycle of the cluster. It references the `AgentClusterInstall` custom resource in the `spec.ClusterInstallRef` setting which defines the cluster parameters. 

Create and apply a `ClusterDeployment` custom resource based on the following example:

[source,yaml]
----
apiVersion: hive.openshift.io/v1
kind: ClusterDeployment
metadata:
  name: single-node
  namespace: demo-worker4
spec:
  baseDomain: hive.example.com
  clusterInstallRef:
    group: extensions.hive.openshift.io
    kind: AgentClusterInstall
    name: test-agent-cluster-install <1>
    version: v1beta1
  clusterName: test-cluster
  controlPlaneConfig:
    servingCertificates: {}
  platform:
    agentBareMetal:
      agentSelector:
        matchLabels:
          location: internal
  pullSecretRef:

>     name: pull-secret

Reviewer 4: ```suggestion
    name: pull-secret <2>
```

The `InfraEnv` custom resource provides the configuration to create the discovery ISO. Within this custom resource, you identify values for proxy settings, ignition overrides, and specifying `NMState` labels. The `spec.nmStateConfigLabelSelector` in this custom resource references the `NMStateConfig` custom reference, so this custom resource must be created after the `NMStateConfig` custom resource if you included one. 

Create and apply the following custom resource:

[source,yaml]
----
apiVersion: agent-install.openshift.io/v1beta1
kind: InfraEnv
metadata:
  name: myinfraenv
  namespace: demo-worker4
spec:
  clusterRef:
    name: single-node  <1>
    namespace: demo-worker4 <2>
  pullSecretRef: 
    name: pull-secret
    sshAuthorizedKey: 'your_public_key_here' <3>
  nmStateConfigLabelSelector:
    matchLabels:
      demo-nmstate-label: value <4>
----

<1> Replace the `clusterDeployment` resource name from <<on-prem-creating-your-cluster-with-the-cli-clusterdeployment,Create the ClusterDeployment>>.

<2> Replace the `clusterDeployment` resource namespace from <<on-prem-creating-your-cluster-with-the-cli-clusterdeployment,Create the ClusterDeployment>>.

> <3> Optional: You can specify your ssh public key, which connects the hub to the machines that were booted from the discovery ISO. 

Reviewer: This just allows the user access to the host when it is booted with the discovery iso.
It doesn't have anything to do with the hub accessing the host as we don't have the private key.

<4> The label name and the label value must match the values in the `label` section of the `NMStateConfig` custom resource that you created in <<on-prem-creating-your-cluster-with-the-cli-nmstateconfig,Optional: Create the NMStateConfig custom resource>>.

[#on-prem-creating-your-cluster-with-the-cli-boot-host]
=== Boot the host from the discovery image

The remaining steps explain how to boot the host from the discovery ISO image that results from the previous procedures. 

. Download the discovery image from the namespace by running the following command:
+
----
curl --insecure -o image.iso $(kubectl -n sample-namespace get infraenvs.agent-install.openshift.io myinfraenv -o=jsonpath="{.status.isoDownloadURL}")
----

. Move the discovery image to virtual media, a USB drive, or another storage location and boot the host from the discovery image that you downloaded. 

. Approve the `Agent` custom resource and start the installation by running the following command:
+
----

> kubectl -n sample-namespace patch agents.agent-install.openshift.io 07e80ea9-200c-4f82-aff4-4932acb773d4 -p '{"spec":{"approved":true}}' --type merge

Reviewer: `oc` rather than `kubectl` maybe?

You can view the status of the installation on the _Clusters_ navigation page.

> Continue with xref:../cluster_lifecycle/access_cluster.adoc#accessing-your-cluster[Accessing your cluster] to access your cluster.

Reviewer 4: I recommend that you italicize the link reference text and move this link to Additional resources. 
```suggestion
Continue with _Accessing your cluster_, which can be found in the _Additional resources_ section, to access your cluster.
```

Or something similar

Author: Leaving this, as it is the end of the procedure and the continuation from that point on.

[#on-prem-creating-your-cluster-with-the-cli]
== Creating your cluster with the command line

> You can also create a single-node cluster without the console by using the assisted installer feature within the Central Infrastructure Management component. After you complete this procedure, you can boot the host from the discovery image that is generated. The order of the procedures is generally not important, but is noted when there is a required order.

Reviewer: Why say "single-node" here. We can create any size cluster.

[#on-prem-creating-your-cluster-with-the-cli-namespace]
=== Create the namespace

> You need a namespace for your resources. It is more convenient to keep all of the resources in a shared namespace. This example uses `sample-namespace` for the name of the namespace, but you can use any name except `assisted-installer`. Create a namepace by creating and applying the following file:

Reviewer 4: ```suggestion
You need a namespace for your resources. It is more convenient to keep all of the resources in a shared namespace. This example uses `sample-namespace` for the name of the namespace, but you can use any name except `assisted-installer`. Create a namespace by creating and applying the following file:
```

Author: This one seems easy to miss. Ha!

[#on-prem-creating-your-cluster-with-the-cli-agentclusterinstall]
=== Create the AgentClusterInstall custom resource

In the `AgentClusterInstall` custom resource, you can specify many of the requirements for the clusters. For example, you can specify the cluster network settings, platform, number of control planes, and worker nodes. 

Create and add the a custom resource that resembles the following example: 

[source,yaml]
----
apiVersion: extensions.hive.openshift.io/v1beta1
kind: AgentClusterInstall
metadata:
  name: test-agent-cluster-install
  namespace: demo-worker4
spec:
  clusterDeploymentRef:
    name: single-node <1>
  imageSetRef:
    name: openshift-v4.12.0 <2>
  networking:
    clusterNetwork:
    - cidr: 10.128.0.0/14
      hostPrefix: 23
    machineNetwork:
    - cidr: 192.168.111.0/24
    serviceNetwork:
    - 172.30.0.0/16
  provisionRequirements:
    controlPlaneAgents: 1
  sshPublicKey: ssh-rsa your-public-key-here <3>
----

<1> Use the same name that you used for your `ClusterDeployment` resource.

<2> Use the `ClusterImageSet` that you generated in <<on-prem-creating-your-cluster-with-the-cli-cluster-image-set,Generate a ClusterImageSet>>.

> <3> Optional: You can include your ssh public key. 

Reviewer: Do you want to add an additional explanation here like you did you infraenv?

Maybe "which allows you to access the host after it is installed"?

> <3> You can specify your ssh public key, which enables you to access the host after it is installed. 

Reviewer 4: ssh should be capital 
```suggestion
<3> You can specify your SSH public key, which enables you to access the host after it is installed. 
```

          address:
            - ip: 192.168.140.30
              prefix-length: 24
          dhcp: false
    dns-resolver:
      config:
        server:
          - 192.168.126.1
    routes:
      config:
        - destination: 0.0.0.0/0
          next-hop-address: 192.168.111.1
          next-hop-interface: eth1
          table-id: 254
        - destination: 0.0.0.0/0
          next-hop-address: 192.168.140.1
          next-hop-interface: eth1
          table-id: 254
  interfaces:
    - name: "eth0"
      macAddress: "02:00:00:80:12:14"
    - name: "eth1"
      macAddress: "02:00:00:80:12:15"
----

<1> The `demo-nmstate-label` label name and its value must be included in the `InfraEnv` resource `spec.nmStateConfigLabelSelector.matchLabels` field.

[#on-prem-creating-your-cluster-with-the-cli-infraenv]
=== Create the InfraEnv custom resource

The `InfraEnv` custom resource provides the configuration to create the discovery ISO. Within this custom resource, you identify values for proxy settings, ignition overrides, and specify `NMState` labels. The value of `spec.nmStateConfigLabelSelector` in this custom resource references the `NMStateConfig` custom resource. 

> *Note:* If you plan to iclude the optional `NMStateConfig` custom resource, the `InfraEnv` custom resource must be created after you create and apply the `NMStateConfig` custom resource. 

Reviewer: ```suggestion
*Note:* If you plan to include the optional `NMStateConfig` custom resource, the `InfraEnv` custom resource must be created after you create and apply the `NMStateConfig` custom resource. 
```

Reviewer: Also I'm pretty sure this isn't true. You should be able to create them in either order and edit the nmstate after the infraenv is created as long as you re-download the ISO after the change.

Author: Made a minor change to the wording, based on the comment.

Create and apply the following custom resource:

[source,yaml]
----
apiVersion: agent-install.openshift.io/v1beta1
kind: InfraEnv
metadata:
  name: myinfraenv
  namespace: demo-worker4
spec:
  clusterRef:
    name: single-node  <1>
    namespace: demo-worker4 <2>
  pullSecretRef: 
    name: pull-secret
    sshAuthorizedKey: 'your_public_key_here' <3>
  nmStateConfigLabelSelector:
    matchLabels:
      demo-nmstate-label: value <4>
----

> <1> Replace the `clusterDeployment` resource name from <<on-prem-creating-your-cluster-with-the-cli-clusterdeployment,Create the ClusterDeployment>>.

Reviewer 4: Instead of using the inline link can you italicize this? Because this is within the file, it doesn't need to go in Additional resources. Link references that direct the user to a different file or folder should be added to the Additional resources section

<3> Optional: You can specify your ssh public key, which enables you to access the host when it is booted from the discovery ISO image. 

<4> The label name and the label value must match the values in the `label` section of the `NMStateConfig` custom resource that you created in <<on-prem-creating-your-cluster-with-the-cli-nmstateconfig,Optional: Create the NMStateConfig custom resource>>.

[#on-prem-creating-your-cluster-with-the-cli-boot-host]
=== Boot the host from the discovery image

The remaining steps explain how to boot the host from the discovery ISO image that results from the previous procedures. 

. Download the discovery image from the namespace by running the following command:

+
----
curl --insecure -o image.iso $(kubectl -n sample-namespace get infraenvs.agent-install.openshift.io myinfraenv -o=jsonpath="{.status.isoDownloadURL}")
----

. Move the discovery image to virtual media, a USB drive, or another storage location and boot the host from the discovery image that you downloaded. 

. Approve the `Agent` custom resource and start the installation by running the following command:

+
----

> oc -n sample-namespace patch agents.agent-install.openshift.io 07e80ea9-200c-4f82-aff4-4932acb773d4 -p '{"spec":{"approved":true}}' --type merge

Reviewer: I think it should be obvious, but do we need to note that this uuid won't always be the same? Or just add a placeholder here like `<agent name>`?

Maybe move the note about how the agent resource is created before this section?

----
+
The `Agent` resource is created automatically. It is registered to the cluster and represents a host that booted from a discovery image. 
+
You can confirm that it was approved when the output of the previous command includes an entry for the target cluster that includes a value of `true` for the `APPROVED` parameter. 

[#additional-resources-cluster-on-premises]
== Additional resources

> - For additional information about how to use zero touch provisioning, see https://access.redhat.com/documentation/en-us/openshift_container_platform/4.13/html/scalability_and_performance/ztp-deploying-disconnected[Clusters at the network far edge] in the {ocp-short} documentation.

Reviewer 4: 2 broken links

```suggestion
- For additional information about how to use zero touch provisioning, see link:https://access.redhat.com/documentation/en-us/openshift_container_platform/4.13/html/scalability_and_performance/ztp-deploying-disconnected[Clusters at the network far edge] in the {ocp-short} documentation.
```

<img width="1028" alt="Screenshot 2023-08-09 at 5 24 24 PM" src="https://github.com/stolostron/rhacm-docs/assets/60616885/511675d2-6ff7-47bb-a3c9-f7570816b308">
